Ffmpeg vce av1 encoder v2

OvchinnikovDmitrii opened the pull request:
Issue: https://github.com/HandBrake/HandBrake/issues/4899

Description of Change:

Adds support for VCE AV1 encoder in HandBrake.

I updated to the latest state of the master branch and made corrections according to the feedback on the previous request:
https://github.com/HandBrake/HandBrake/pull/4900
I created a new pull request due to the complexity of the rebase in order to avoid errors.

**Tested on:**

- [x] Windows 10+  (via MinGW)
- [ ] macOS 10.13+
- [ ] Ubuntu Linux

galad87 committed:
libhb: generalize the hwaccel implementation.
Introduce a way to wrap AVFrame and CMSampleBuffer inside a hb_buffer_t to avoid unneeded conversions. The wrapping is enabled only for hw AVFrame at the moment.
Enable VideoToolbox hw decoders.

Nomis101 committed:
[MacGUI] Update all locales

OvchinnikovDmitrii committed:
libhb: FFmpeg VCE AV1 encoder registration.

OvchinnikovDmitrii committed:
libhb: FFmpeg VCE AV1 encoder registration.

OvchinnikovDmitrii committed:
preset: add AV1 VCN 2160p 4K preset

OvchinnikovDmitrii committed:
preset: add AV1 VCN 2160p 4K preset

galad87 commented:
Is 10 bit not supported, or is it always 10 bit?
It seems a couple of tabs sneaked in, can you replace them with spaces?

sr55 committed:
WinGui: Fix an issue with the Queue Stop Button not being available when it needs to be. Fixes #5165

sr55 committed:
WinGui: Fix Preset move control menu visibility

sr55 committed:
WinGui: Fix a bug on queue import that could result in Force/Burn-in not being selectable or selected correctly. Fixes #5163

OvchinnikovDmitrii committed:
libhb: encavcodec - replace tabs with spaces

OvchinnikovDmitrii committed:
libhb: encavcodec - replace tabs with spaces

OvchinnikovDmitrii commented:
> Is 10 bit not supported, or is it always 10 bit? It seems a couple of tabs sneaked in, can you replace them with spaces?

I replaced tabs with spaces. I plan to add 10 bits for AV1 later.

galad87 commented on libhb/muxavformat.c:
```diff
@@ -438,7 +438,24 @@ static int avformatInit( hb_mux_object_t * m )
                 }
             }
         } break;
+        case HB_VCODEC_FFMPEG_VCE_AV1:
```
This code looks like a duplicate of the existing one for HB_VCODEC_SVT_AV1 and HB_VCODEC_FFMPEG_NVENC_AV1 a few lines above.

galad87 commented:
I can't test it, but apart from the comment above it looks good.

bradleysepos commented:
I will test when I can make some time, possibly this weekend.

galad87 committed:
MacGui: expand the automatic file naming options.

galad87 committed:
MacGui: add an option to use the ISO date format for autonamin.

galad87 committed:
MacGui: add 'Set destination to source folder' option in the preferences. Under the sandbox it will ask the user for write access if the destination is different than the previous choosen one.

galad87 committed:
MacGui: don't check for the writable status when setting the destination to be the same as source, because it doesn't actually work under the sandbox.

galad87 committed:
MacGui: move the same as source destination option to its own section.

galad87 committed:
Cosmetics.

galad87 committed:
MacGui: improve management of security scoped resources lifetime from URL passed from open panels.

galad87 committed:
MacGui: remove some unneeded code, we always do a title specific scan when reloading a job from queue.

galad87 committed:
contrib: change the 3gpp udta reader to only warn about malformed tags instead of failing opening the whole file.

sr55 committed:
VCN: Be explicit in setting "main10" profile for the HEVC 10bit encoder. There appears to have been a regression in the AMD driver that now incorrectly sets this if left to choose itself.  #5169

galad87 committed:
libhb: iconv functions return an iconv_t, the additional * is unneeded.

galad87 committed:
cli: fix the --audio-lang-list documentation. 'any' is used to match all languages since ed346b4.

sr55 committed:
WinGui: Add support for ass files on drag/drop and fix a crash condition when importing subtitle files.

sr55 commented:
Shall we push and let folks in the userbase have a go at it in the mean time?

bradleysepos commented:
Got sidetracked, give me a few days.

Nomis101 committed:
Add all DTS profiles to scan.c

Fixes #5186

bradleysepos commented:
I made some time to look at this. Thanks for rebasing on latest master branch. A few questions:

1. It would be good to understand what is going on with https://trac.ffmpeg.org/ticket/10266. I found this in my testing of the previous PR after giving feedback, and have also experienced the issue with OBS Studio. If a user sets the resolution limit to 1080p in HandBrake and the encoder outputs 1920x1082, then the resulting file will not be playable on devices only supporting 1080p, which is a problem. Any info you have on this and whether it is something that can be fixed is appreciated.
2. Same question as on the v1 PR: CQ in the GUI ranges from 1-31. Are there CQ values higher than 31 (would make sense)?
3. Same question as on the v1 PR: MPV was able to play my test encode, but seeking did not work. To be more specific, audio plays but video is a black screen when seeking. Perhaps an issue with (missing) IDR frames?
4. The tabs to spaces commit should be squashed (fixup) into the first commit. There are some tabs in the preset .json that need addressed also.
5. galad87's comment regarding code duplication could be addressed. The rest of the code looks good to me.

brad0 committed:
contrib: update libdav1d to 1.2.1.

sr55 reviewed (commented):
hb_video_quality_get_limits will also need updated, otherwise the UI will show the range as 1-31 which limits the low range on the quality output

galad87 committed:
qsv: remove some unused functions and structs and old workarounds.

galad87 committed:
qsv: disable the gpu full path if the source is rotated.

brad0 committed:
contrib: update to libxml2 2.11.4

galad87 committed:
libhb: add VideoToolbox HW rotate and cropscale filters.

galad87 committed:
contrib: update SVT-AV1 to version 1.6.0.

galad87 committed:
contrib: do not use the --offline option for libdovi, it seems to make the build randomly fail on GitHub CI.

galad87 committed:
build: link with ntdll on mingw.

galad87 committed:
Revert "MacGui: remove the '…' at the end of some strings. It could give the impression that the whole string is not displayed. #5100."
Actually it's common to use '…' to indicate a long running operation.

This reverts commit 9341c06e4dea27ef95ff1a410e0676e29123ff53.

galad87 committed:
libhb: fail early when the colorspace is enabled but the color profile is not set. It can happens when the source is dolby vision profile 5, and the colorspace filter doesn't know how to handle it.

OvchinnikovDmitrii committed:
libhb: encavcodec - cleanup

OvchinnikovDmitrii committed:
libhb: encavcodec - cleanup

twalker314 committed:
build: use documented way of indicating whitespace in regexp for grep

Fixes #5206

Also remove redundant escapes before dashes causing
GNU grep to print warnings about "stray \ before -"

galad87 committed:
libhb: fix svt-av1 psnr and fastdecode options.

galad87 committed:
libhb: add more biplanar pixel formats.

galad87 committed:
libhb: add a way to force the usage of the hw decoder.

sr55 committed:
Add support for excluded File extensions (#5218)

libhb: Add support for excluding file extensions for batch scan. (#2541)
Consolidated hb_scan and hb_scan2
Add a new preference to the Windows UI to allow controlling this option. By Default: png, jpg, txt, srt, ssa, ass are excluded. Users can change this if they choose.

sr55 committed:
WinGui: Handle a possible null reference issue.

sr55 committed:
WinGui: Add underlying support for scans being run in a remote process. This is not yet wired up.

sr55 committed:
WinGui: Some tidyup / Refactoring of the code post last 2 commits.

Nomis101 committed:
Update libjpeg-turbo to 3.0.0

ghost commented:
just passing by, thought i'd mention this for posterity
> Same question as on the v1 PR: MPV was able to play my test encode, but seeking did not work. To be more specific, audio plays but video is a black screen when seeking. Perhaps an issue with (missing) IDR frames?

it's funny i get this issue on intel graphics https://trac.ffmpeg.org/ticket/10283#comment:1

galad87 committed:
libhb: add missing defines and improve hw filters validation.

galad87 committed:
MacGui: fix a leak of security scoped resources that happens when importing external subtitles.

galad87 committed:
MacGui: call stopAccessingSecurityScopedResource on the remaining URLs from the open panels.

OvchinnikovDmitrii committed:
libhb: encavcodec

libhb: encavcodec - fix quality limits for vce av1

OvchinnikovDmitrii committed:
libhb: encavcodec

libhb: encavcodec - fix quality limits for vce av1

galad87 committed:
libhb: correctly check for the hw_decode if the force hw decode option is set.

galad87 committed:
libhb: disable HDR10+ passthru when the encoder is vt_h265_10bit. VideoToolbox doesn't preserve such metadata yet.

galad87 committed:
LinGui: enable the video option extra for nvenc encoders.

Nomis101 committed:
Update zimg to 3.0.5

bradleysepos commented:
It looks like another AMD developer has confirmed (linked via the [aforementioned FFmpeg ticket](https://github.com/HandBrake/HandBrake/pull/5164#issuecomment-1586535593)) the 1920x1082 issue is a hardware limitation of VCN 4 and won't be fixed until VCN 5. We should probably consider the AV1 encoder experimental at this stage, as it cannot produce the requested output. We'll still merge this PR when it is finished, but I wonder whether we should postpone adding the preset and let people select the encoder manually. Comments, anyone?

Looks like the [seeking issue (no. 3)](https://github.com/HandBrake/HandBrake/pull/5164#issuecomment-1586535593) is what remains to address and/or discuss.

Nomis101 committed:
[WinGUI] Update all 100% locale files (#5249)

sr55 committed:
File Scan Improvements (#5236)

- Add an hb_scan_list method that takes a list of files to scan. This allows a user to drag any arbitrary list of files onto the UI.
- Add a check for known file types during scan. When a known file type is detected, bypass the DVD / Bluray Probe. This reduces some log spam.
- Supported in Mac and Windows UIs

---------

Co-authored-by: Damiano Galassi

sr55 committed:
Couple updates to NEWS

sr55 committed:
WinGui: Add support for recursive folder scan. This can be enabled in Preferences -> Advanced.   #1310

sr55 committed:
NEWS updates

galad87 committed:
MacGui: add support for recursive folder scan. Can be enabled in the open panel.

galad87 committed:
MacGui: fix the destination same as source mode after the multi files selection improvements.

galad87 commented:
http://ffmpeg.org/pipermail/ffmpeg-devel/2023-July/311961.html might help the 3)

galad87 committed:
MacGui: improve the finding of a base common url for selected urls.

sr55 committed:
WinGui: More robust error handling when probing progress of the remote worker. Should fix #5251

galad87 committed:
libhb: use the r_frame_rate avstream field if available and nb_frames is not set.

galad87 committed:
cli: always use at least 1 preview.

galad87 committed:
libhb: do not set a maximum bitrate for vp9, it's not needed.

galad87 committed:
libhb: avoid a log about interlaced title if there is only one preview image and it's not interlaced.

OvchinnikovDmitrii committed:
Merge branch 'FFmpeg_VCE_AV1_encoder_v2' of https://github.com/amfdev/HandBrake into FFmpeg_VCE_AV1_encoder_v2

# Conflicts:
#	libhb/encavcodec.c

OvchinnikovDmitrii committed:
libhb: encavcodec - fix include and seeking

OvchinnikovDmitrii commented:
> http://ffmpeg.org/pipermail/ffmpeg-devel/2023-July/311961.html might help the 3)

I added this patch and rebase to current master. Now seeking works fine.

bradleysepos commented:
> > http://ffmpeg.org/pipermail/ffmpeg-devel/2023-July/311961.html might help the 3)
> 
> I added this patch and rebase to current master. Now seeking works fine.

Verified seeking works with the patch, so all addressable issues have been dealt with.

I've rebased this branch and added one additional commit as part of #5304. Let's continue discussion there.
